runtime/lib/capi/Driver.hpp

```diff
@@ -0,0 +1,141 @@
+// Copyright 2022-2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cassert>
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "QuantumDevice.hpp"
+
+#if __has_include("StateVectorCPU.hpp")
+#include "LightningSimulator.hpp"
+#endif
+
+#if __has_include("StateVectorKokkos.hpp")
+#include "LightningKokkosSimulator.hpp"
+#endif
+
+namespace Catalyst::Runtime::CAPI {
+
+class MemoryManager final {
+  private:
+    std::unordered_set<void *> _impl;
+
+  public:
+    explicit MemoryManager() { _impl.reserve(1024); };
+
+    ~MemoryManager()
+    {
+        for (auto allocation : _impl) {
+            free(allocation);
+        }
+    }
+
+    void insert(void *ptr) { _impl.insert(ptr); }
+    void erase(void *ptr) { _impl.erase(ptr); }
+};
+
+class Driver final {
+  private:
+    using DeviceInitializer = std::function<std::unique_ptr<QuantumDevice>(bool, size_t)>;
+    // TODO: remove this and use `QuantumDeviceMap` after removing `CreateQuantumDevice`
+    std::unordered_map<std::string_view, DeviceInitializer> _device_map{
+#ifdef __device_lightning
+        {"lightning.qubit",
+         [](bool tape_recording, size_t shots) {
+             return std::make_unique<Catalyst::Runtime::Simulator::LightningSimulator>(
+                 tape_recording, shots);
+         }},
+#endif
+#ifdef __device_lightning_kokkos
+        {"lightning.kokkos",
+         [](bool tape_recording, size_t shots) {
+             return std::make_unique<Catalyst::Runtime::Simulator::LightningKokkosSimulator>(
+                 tape_recording, shots);
+         }},
+#endif
```

Comment:
These ifdefs will be removed in the future, what is the idea behind the future implementation. And will it be part of this PR or another?

Reply:
Ah! I thought we discuss it offline; the focus of this PR is not the backend part (check PR #89). 

Within the scope of this PR, we still cannot compile and link against both simulators. I'm temporary using preprocesser directives for each backend simulator to check which one is compiled but they are temporary and there shouldn't be required if we always build these backend simulators in the runtime.